File: src/models/Routine.js
import { Cadence, Category, DeliveryMethod } from ".";

class Routine {
  static default_limit = 3;
  #_id;
  #_name;
  #_category;
  #_delivery_method;
  #_cadence;
  #_limit;
  #_created;
  #_start_date;
  #_next_date;
  #_iterations;

  constructor(name) {
    this.#_name = name;
    this.#_created = new Date();
    this.#_category = Category.Life;
    this.#_iterations = 0;
    this.#_limit = Routine.default_limit;
    this.#_delivery_method = DeliveryMethod.Email;
    this.#_cadence = Cadence.Weekly;
    this.#_start_date = this.#_cadence._start_date;
    this.#_next_date = this.#_cadence.next_date;
  }

  // METHODS
  increment_iterations() {
    this.#_iterations++;
  }

  reset_iterations() {
    this.#_iterations = 0;
    this.#_start_date = new Date();
  }

  reset_cadence() {
    this.#_cadence = Cadence.Weekly;
  }

  snooze(freq, measure_of_time) {
    this.#_cadence.snooze(freq, measure_of_time);
  }

  change_category(category) {
    this.#_category = category;
  }

  // SETTERS
  set id(id) {
    this.#_id = id;
  }
  set name(name) {
    this.#_name = name;
  }
  set category(category) {
    this.#_category = category;
  }
  set delivery_method(delivery_method) {
    this.#_delivery_method = delivery_method;
  }
  set cadence(cadence) {
    this.#_cadence = cadence;
  }
  set limit(limit) {
    this.#_limit = Math.min(limit, Routine.limit);
  }
  set start_date(start_date) {
    this.#_start_date = start_date;
  }
  set next_date(next_date) {
    this.#_next_date = next_date;
  }

  // GETTERS
  get name() {
    return this.#_name;
  }

  get category() {
    return this.#_category;
  }

  get delivery_method() {
    return this.#_delivery_method;
  }

  get cadence() {
    return this.#_cadence;
  }

  get limit() {
    return this.#_limit;
  }

  get start_date() {
    return this.#_start_date;
  }

  get next_date() {
    return this.#_next_date;
  }

  get iterations() {
    return this.#_iterations;
  }

  get created() {
    return this.#_created;
  }
}

export default Routine;

File: data/routine.py
from datetime import datetime, timedelta
from operator import itemgetter
from .delivery_method import DeliveryMethod


class Routine:
    today = datetime.now()
    Cadence = None

    def __init__(self, routine, Cadence) -> None:
        Routine.Cadence = Cadence

        id, name, category, cadence, deliveryMethod, lastDate, nextDate, iterations, created = itemgetter(
            'id', 'name', 'category', 'cadence', 'deliveryMethod', 'lastDate', 'nextDate', 'iterations', 'created')(routine)

        self.id = id
        self.name = name
        self.category = category
        self.iterations = iterations
        self.deliveryMethod = deliveryMethod
        self.cadence = Cadence[cadence]
        self.created = datetime.strptime(created, '%m/%d/%Y')
        self.lastDate = datetime.strptime(
            lastDate, '%m/%d/%Y')
        self.nextDate = datetime.strptime(nextDate, '%m/%d/%Y')

    def update(self, contact_info={}, delay=None):
        today = Routine.today.date()
        upcoming = self.nextDate.date()

        sendReminder = (upcoming - today).days == self.cadence.value[1]

        if delay:
            self.nextDate = today + timedelta(days=self.cadence.value[1])
            return self
        elif sendReminder:
            self.remind(contact_info)
        elif upcoming == today:
            self.lastDate = today
            self.nextDate = today + timedelta(days=self.cadence.value[0])
            return self
        return False

    def remind(self, contact_info):
        DeliveryMethod(contact_info, self)

    def increment_iterations(self):
        self.iterations += 1

    def reset_iterations(self):
        self.iterations = 0
        self.lastDate = datetime.now()

    def reset_cadence(self):
        self.cadence = Routine.Cadence((1, 1)).name

    def snooze(self):
        self.update({}, True)

    def change_category(self, category):
        self.category = category
